File: Scripts/heatmap.py
# Annotation_Information_Start --------------------------------------------

# Author: scaddleloong@gmail.com
# Date: 2026/06/03, 07: 26, PM
# Create Heatmap.html

# Annotation_Information_End ----------------------------------------------

import datetime as dt
import plotly.graph_objects as go
from database import load_all

current_year = dt.date.today().year
HEATMAP_PATH = f"../Output/Daily_Heatmap-{current_year}.html"

def generate_heatmap():

    rows = load_all()
    if not rows:
        return

    today = dt.date.today()
    current_year = today.year
    start_date = dt.date(current_year, 1, 1)
    end_date = dt.date(current_year, 12, 31)

    data = {
        dt.datetime.strptime(d, "%Y-%m-%d").date(): s / 60
        for d, s in rows
    }

    anchor = start_date - dt.timedelta(days=start_date.weekday())

    total_days = (end_date - anchor).days + 1
    total_weeks = (total_days // 7) + 1

    z = [[0 for _ in range(total_weeks)] for _ in range(7)]
    text = [["" for _ in range(total_weeks)] for _ in range(7)]

    current = start_date
    while current <= end_date:

        day_index = (current - anchor).days
        week_index = day_index // 7
        weekday = current.weekday()

        minutes = round(data.get(current, 0), 1)

        z[weekday][week_index] = minutes
        text[weekday][week_index] = (
            f"{current}<br>"
            f"Reading Time: {minutes:.1f} minutes"
        )
        current += dt.timedelta(days=1)

    month_positions = []
    month_labels = []

    current = start_date
    while current <= end_date:
        if current.day == 1:
            week_index = (current - anchor).days // 7
            month_positions.append(week_index)
            month_labels.append(current.strftime("%Y-%m"))
        current += dt.timedelta(days=1)

    fig = go.Figure(
        data=go.Heatmap(
            z=z,
            text=text,
            hoverinfo="text",
            xgap=2,
            ygap=2,

            zmin=0,
            zmax=240,

            colorscale=[
                [0.00, "#ebedf0"],
                [0.20, "#9be9a8"],
                [0.40, "#40c463"],
                [0.60, "#30a14e"],
                [0.80, "#216e39"],
                [1.00, "#0e4429"]
            ],

            showscale=False,
        )
    )

    fig.update_layout(
        title="Zotero Reading Record (2026)",
        height=300,
        width=max(1400, total_weeks * 25),

        plot_bgcolor="white",
        paper_bgcolor="white",

        margin=dict(l=60, r=30, t=70, b=40),

        yaxis=dict(
            tickmode="array",
            tickvals=list(range(7)),
            ticktext=["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"],
            autorange="reversed",
            showgrid=False,
            zeroline=False
        ),

        xaxis=dict(
            tickmode="array",
            tickvals=month_positions,
            ticktext=month_labels,
            side="top",
            showgrid=False,
            zeroline=False
        )
    )

    fig.write_html(HEATMAP_PATH, include_plotlyjs="cdn")

    print(f"[{dt.datetime.now()}] Heatmap Updated: {HEATMAP_PATH}")

File: Plugin_Turning/V0.3/heatmap.js
/* Annotation_Information_Start

Author: scaddleloong@gmail.com
Date: 2026/06/09 11: 17 AM
heatmap.js V0.3

Annotation_Information_End */

// generate heatmap function
function generateHeatmapHTML() {

    const rows = loadAll(); // function from storage

    const year = new Date().getFullYear();

    const startDate = new Date(year, 0, 1);
    const endDate = new Date(year, 11, 31);

    const map = {};

    for (const r of rows) {
        map[r.date] = r.seconds / 60; // change to minutes
    }

    // statistics
    const totalMinutes =
        rows.reduce((sum, r) => sum + r.seconds / 60, 0);

    const totalHours =
        (totalMinutes / 60).toFixed(1);

    const activeDays =
        rows.filter(r => r.seconds > 0).length;

    const avgMinutes =
        activeDays > 0
            ? (totalMinutes / activeDays).toFixed(1)
            : "0.0";

    const anchor = new Date(startDate);

    // start from monday
    anchor.setDate(
        anchor.getDate() -
        ((anchor.getDay() + 6) % 7)
    );

    const cells = [];

    let cur = new Date(anchor);

    while (cur <= endDate) {

        const iso =
            cur.getFullYear() +
            "-" +
            String(cur.getMonth() + 1).padStart(2, "0") +
            "-" +
            String(cur.getDate()).padStart(2, "0");

        const minutes = map[iso] || 0;

        let level = 0;

        // color levels
        if (minutes > 15) level = 1;
        if (minutes >= 30) level = 2;
        if (minutes >= 60) level = 3;
        if (minutes >= 120) level = 4;

        cells.push(`
            <div
                class="cell level-${level}"
                title="${iso} | ${minutes.toFixed(1)} min"
            ></div>
        `);

        cur.setDate(cur.getDate() + 1);
    }

    return `
<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">

<title>Reading Heatmap</title>

<style>

html,
body{
    background:white;
}

body{
    font-family:Arial,sans-serif;
    margin:20px;
    color:#24292f;
}

h2{
    margin-bottom:10px;
}

.stats{
    margin-bottom:20px;
    font-size:14px;
}

.stats div{
    margin-top:4px;
}

.heatmap-wrapper{
    display:flex;
    align-items:flex-start;
}

.weekday-labels{

    display:grid;

    grid-template-rows:
        repeat(7,16px);

    gap:4px;

    margin-right:10px;

    font-size:11px;

    color:#666;

    width:30px;
}

#heatmap{

    display:grid;

    grid-auto-flow:column;

    grid-template-rows:
        repeat(7,16px);

    grid-auto-columns:16px;

    gap:4px;

    width:max-content;
}

.cell{
    width:16px;
    height:16px;

    border-radius:3px;

    background:#ebedf0;
}

.level-0{
    background:#ebedf0;
}

.level-1{
    background:#9be9a8;
}

.level-2{
    background:#40c463;
}

.level-3{
    background:#30a14e;
}

.level-4{
    background:#216e39;
}

#months{

    display:grid;

    grid-template-columns:
        repeat(12,1fr);

    width:1060px;

    margin-top:10px;

    margin-left:40px;

    font-size:11px;

    color:#666;
}

.legend{
    margin-top:20px;

    display:flex;
    align-items:center;

    gap:5px;

    font-size:12px;
}

.legend-box{
    width:16px;
    height:16px;
    border-radius:3px;
}

</style>
</head>

<body>

<div class="stats">

    <h2>Reading Statistics (${year})</h2>

    <div>Total Reading Time: ${totalHours} h</div>

    <div>Active Days: ${activeDays}</div>

    <div>Average: ${avgMinutes} min/day</div>

</div>

<div class="heatmap-wrapper">

    <div class="weekday-labels">
        <div>Mon</div>
        <div>Tue</div>
        <div>Wed</div>
        <div>Thu</div>
        <div>Fri</div>
        <div>Sat</div>
        <div>Sun</div>
    </div>

    <div id="heatmap">
        ${cells.join("")}
    </div>

</div>

<div id="months">

    <span>Jan</span>
    <span>Feb</span>
    <span>Mar</span>
    <span>Apr</span>
    <span>May</span>
    <span>Jun</span>
    <span>Jul</span>
    <span>Aug</span>
    <span>Sep</span>
    <span>Oct</span>
    <span>Nov</span>
    <span>Dec</span>

</div>

<div class="legend">

    <span>Less</span>

    <div class="legend-box level-0"></div>
    <div class="legend-box level-1"></div>
    <div class="legend-box level-2"></div>
    <div class="legend-box level-3"></div>
    <div class="legend-box level-4"></div>

    <span>More</span>

</div>

<p style="margin-top:15px;font-size:12px;color:#666;">
Hover over a square to view reading time.
</p>

</body>
</html>
`;
}

// open heatmap window in new window
function openHeatmapWindow() {

    const html = generateHeatmapHTML();

    const win = Services.ww.openWindow(
    null,
    "about:blank",
    "_blank",
    "chrome,resizable,scrollbars,width=1200,height=800,left=100,top=100",
    null
);

win.addEventListener("load", () => {

    win.resizeTo(1200, 500);
    win.moveTo(100, 200);

    win.document.open();
    win.document.write(html);
    win.document.close();
});
}